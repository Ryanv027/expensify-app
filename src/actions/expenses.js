import database from './../firebase/firebase';

export const addExpense = (expense)  => ({
    type: 'ADD_EXPENSE',
    expense
  });
  
  export const removeExpense = ({id = ''} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
  });
  
  export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
  })
  
  export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
      } = expenseData;
      const expense = { description, note, amount, createdAt };
      return database.ref('expenses').push(expense).then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })
    };
  };
  
  //===================================================================================================================================
  
  
  //OPTION 1
  //component calls action generator
  // action generator returns object
  // component dispatches object
  // redux store changes 
  
  //OPTION 2
  //component calls action generator
  // action generator returns function
  // component dispatches function (?)
  // redux store changes (has the ability to dispatch other actions and do whatever it wants)
