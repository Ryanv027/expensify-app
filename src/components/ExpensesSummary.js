import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getSelectedExpenses from './../selectors/expenses-total';
import selectExpenses from './../selectors/expenses';

export const ExpensesSummary = (props) => {
  const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(props.expensesTotal / 100).format('$0,0.00')
  return (
    <div>
      <p>{`Viewing ${props.expenseCount} ${expenseWord} totalling ${formattedExpensesTotal}`}</p>
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenseCount: selectExpenses(state.expenses, state.filters).length,
  expensesTotal: getSelectedExpenses(selectExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);