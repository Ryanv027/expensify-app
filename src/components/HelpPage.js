import React from 'react';

const HelpPage = () => (
    <div>This is From the Help Page</div>
);

export default HelpPage;