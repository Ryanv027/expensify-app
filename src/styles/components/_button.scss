.button {
  background: $light-blue;
  border: none;
  color:white; 
  display: inline-block;
  font-weight: 300;
  font-size: $font-size-large;
  line-height: 1;
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none; 
}

.button--remove{
  background: rgb(146, 0, 0);
  margin-bottom: 15px;
}