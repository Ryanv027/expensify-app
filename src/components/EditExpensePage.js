import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from './../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    const confirmation = confirm('Are you sure you want to delete this expense?');
    if(confirmation === true){
      const id = this.props.expense.id;
      this.props.startRemoveExpense(id);
      this.props.history.push('/');
    } else return false;
  };
  render(){
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Edit Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm 
          onSubmit={this.onSubmit}
          expense={this.props.expense}
          />
          <button
            className='button button--remove'
            onClick={this.onRemove}
            >Remove Expense
          </button>
        </div>
      </div>
    )
  };
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);