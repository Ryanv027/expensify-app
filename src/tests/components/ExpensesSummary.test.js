import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from './../../components/ExpensesSummary';
import getSelectedExpenses from './../../selectors/expenses-total';
import expenses from './../fixtures/expenses';

test('should render ExpensesSummary correctly with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary />);
  wrapper.setProps({
    expenseCount: 2,
    expensesTotal: getSelectedExpenses([expenses[0], expenses[1]])
  });
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly a single expense', () => {
  const wrapper = shallow(<ExpensesSummary />);
  wrapper.setProps({
    expenseCount: 1,
    expensesTotal: getSelectedExpenses([expenses[1]])
  });
  expect(wrapper).toMatchSnapshot();
});