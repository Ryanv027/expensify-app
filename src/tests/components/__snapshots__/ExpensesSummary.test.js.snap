// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly a single expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $75.00
  </p>
</div>
`;

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<div>
  <p>
    Viewing 2 expenses totalling $76.95
  </p>
</div>
`;
